### PR TITLE
Calendar

### PR DESCRIPTION
**Pull request by radavis**

### CODE REVIEW

File: app/models/calendar.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
+require "google/api_client"
+
+class Calendar < ActiveRecord::Base
```
Comment (dpickett) [inline app/models/calendar.rb]: how do I correlate the calendar with a team?

File: app/models/calendar.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  has_many :calendar_events
+  has_many :events, class_name: CalendarEvent
```
Comment (dpickett) [inline app/models/calendar.rb]: `class_name` requires a stringified value

File: app/models/calendar.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    response = client.execute(
+      api_method: client.discovered_api("calendar", "v3").events.list,
+      parameters: { "calendarId" => cid }
+    )
+
+    JSON.parse(response.body)
+  end
+
+  def import_events
+    calendar_json["items"].each do |event_json|
```
Comment (dpickett) [inline app/models/calendar.rb]: what happens if a calendar event is deleted from a calendar?

File: app/models/calendar.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      event_params = {
+        title: event_json["summary"],
+        from: event_json["start"].try(:[], "dateTime") ||
+          event_json["start"].try(:[], "date"),
+        to: event_json["end"].try(:[], "dateTime") ||
+          event_json["end"].try(:[], "date"),
+        url: event_json["htmlLink"]
+      }
+
+      event = CalendarEvent.find_or_initialize_by(id: event_json["id"])
```
Comment (dpickett) [inline app/models/calendar.rb]: If we're going to persist the calendar event separately, I would recommend we store the external calendar id somewhere other than the primary key.

File: db/structure.sql [hunk 5]
```diff
@@ -97,6 +97,74 @@ ALTER SEQUENCE assignments_id_seq OWNED BY assignments.id;
...
+    id integer NOT NULL,
+    title character varying(255) NOT NULL,
+    "from" timestamp without time zone NOT NULL,
+    "to" timestamp without time zone NOT NULL,
+    url text,
+    eid character varying(255),
+    calendar_id integer,
+    created_at timestamp without time zone,
+    updated_at timestamp without time zone
+);
```
Comment (radavis) [inline db/structure.sql]: We should get rid of this table

File: db/structure.sql [hunk 6]
```diff
@@ -97,6 +97,74 @@ ALTER SEQUENCE assignments_id_seq OWNED BY assignments.id;
...
+
+
+--
+-- Name: calendars; Type: TABLE; Schema: public; Owner: -; Tablespace: 
+--
+
+CREATE TABLE calendars (
+    id integer NOT NULL,
+    name character varying(255) NOT NULL,
+    cid character varying(255) NOT NULL,
```
Comment (radavis) [inline db/structure.sql]: Add uniqueness index on cid as well as model validations.

File: app/models/calendar.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  validates :cid, presence: true, uniqueness: true
+
+  def events_json
+    redis = Redis.new
+    result = redis.get(cid)
+    if result
+      result = JSON.parse(result)
+    else
+      result = GoogleCalendarAdapter.new(cid)
+        .fetch_events(default_start_time, default_end_time)
```
Comment (radavis) [inline app/models/calendar.rb]: Nope.

File: Gemfile [hunk 8]
```diff
@@ -32,11 +34,13 @@ group :development, :test do
   gem "capybara"
   gem "factory_girl_rails"
   gem "pry-rails"
+  gem "shoulda-matchers"
 end
 
 group :test do
   gem "coveralls", require: false
   gem "launchy", require: false
+  gem "database_cleaner"
```
Comment (atsheehan) [inline Gemfile]: Is there a reason to use `database_cleaner` rather than the transactional fixtures in rspec?

Comment (dpickett) [inline Gemfile]: in the past, I've found transactional fixtures to not behave deterministically. I've found `database_cleaner` to behave more reliably.

File: spec/features/view_dashboard_calendar_spec.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+  scenario "user sees event information" do
+    sign_in_as(user)
+    visit dashboard_path
+    expect(page).to have_content("Monday, February 9 at 19:00")
+    expect(page).to have_link("Community: Boston MySQL Monthly Meetup")
+  end
+
+  scenario "events that have already started have a class of '.past-event'" do
+    pending "override Calendar default_start_time default_end_time methods"
```
Comment (radavis) [inline spec/features/view_dashboard_calendar_spec.rb]: Delete this line

File: spec/lib/google_calendar_adapter_spec.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+      events = calendar.fetch_events
+      expect(events).to be_an Array
+      expect(events.first["kind"]).to eq("calendar#event")
+    end
+
+    it "should scope events to a date range" do
+      start_time = DateTime.parse("2015/01/26").beginning_of_day
+      end_time = DateTime.parse("2015/01/30").end_of_day
+      events = calendar.fetch_events(start_time, end_time)
+      event_time = DateTime.parse events.first["start"]["dateTime"]
```
Comment (radavis) [inline spec/lib/google_calendar_adapter_spec.rb]: event_time = DateTime.parse(events.first["start"]["dateTime"])

File: app/models/calendar.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    redis.set(cid, result.to_json)
+    redis.expire(cid, 15.minutes)
+
+    result
+  end
+
+  def events
+    results = []
+    events_json.each do |event_json|
+      results << CalendarEvent.new(event_json)
```
Comment (dreamingfurther) [inline app/models/calendar.rb]: This implementation for creating a CalendarEvent is certainly functional but one thing it does is leak the structure of the redis data into our own code. 

What do I mean by this? Basically by passing whatever JSON is returned into the CalendarEvent.new in a hash the CalendarEvent object needs to care/know about the format of the data coming out of the redis database.

One way we could clean up the current structure to keep the knowledge of the redis JSON data out of the CalendarEvent object would be to create a hash that is passed into the initialize method of the CalendarEvent. That way if the structure of the Redis JSON data changes we won't have to change the CalendarEvent object itself at all, we'll just need to update whatever code is creating the hash that is passed into its initialize method.

I'd be happy to pair on a small refactor to clean this up.

**Pull request closed without merging**